vim-patch:8.1.1542,8.2.3549: an OptionSet autocommand does not get enough info

zeertzjq opened the pull request:
Close #9649

#### vim-patch:8.1.1542: an OptionSet autocommand does not get enough info

Problem:    An OptionSet autocommand does not get enough info.
Solution:   Add v:option_command, v:option_oldlocal and v:option_oldglobal.
            (Latrice Wilgus, closes vim/vim#4118)
https://github.com/vim/vim/commit/d7c968794710f338d491072171df48f96612cf72

#### vim-patch:8.2.3549: mistakes in test comments
    
Problem:    Mistakes in test comments.
Solution:   Fix the comments. (closes vim/vim#9029)
https://github.com/vim/vim/commit/b811de5d49a23b20df82a80533d0602cda49029e

janlazo commented:
`DICTITEM16_KEY_LEN 16` macro is missing. It's used in `eval_init()`. 
`dictitem16_S` was replaced with ChangedtickDictItem.

janlazo commented on src/nvim/eval.c:
```diff
@@ -10450,8 +10453,11 @@ void option_last_set_msg(LastSet last_set)
 // reset v:option_new, v:option_old and v:option_type
 void reset_v_option_vars(void)
 {
```
Function comment wasn't ported.

zeertzjq commented:
> `DICTITEM16_KEY_LEN 16` macro is missing. It's used in `eval_init()`. `dictitem16_S` was replaced with ChangedtickDictItem.

I don't think that macro has anything to do with `ChangedtickDictItem`.

janlazo commented:
```c
/*
 * Initialize the global and v: variables.
 */
void eval_init(void)
{
  vimvars[VV_VERSION].vv_nr = VIM_VERSION_100;

  struct vimvar *p;

  init_var_dict(&globvardict, &globvars_var, VAR_DEF_SCOPE);
  init_var_dict(&vimvardict, &vimvars_var, VAR_SCOPE);
  vimvardict.dv_lock = VAR_FIXED;
  hash_init(&compat_hashtab);
  func_init();

  for (size_t i = 0; i < ARRAY_SIZE(vimvars); i++) {
    p = &vimvars[i];
    assert(STRLEN(p->vv_name) <= 16);
    STRCPY(p->vv_di.di_key, p->vv_name);
    if (p->vv_flags & VV_RO) {
      p->vv_di.di_flags = DI_FLAGS_RO | DI_FLAGS_FIX;
    } else if (p->vv_flags & VV_RO_SBX) {
      p->vv_di.di_flags = DI_FLAGS_RO_SBX | DI_FLAGS_FIX;
    } else {
      p->vv_di.di_flags = DI_FLAGS_FIX;
    }

    // add to v: scope dict, unless the value is not always available
    if (p->vv_type != VAR_UNKNOWN) {
      hash_add(&vimvarht, p->vv_di.di_key);
    }
    if (p->vv_flags & VV_COMPAT) {
      // add to compat scope dict
      hash_add(&compat_hashtab, p->vv_di.di_key);
    }
  }
  vimvars[VV_VERSION].vv_nr = VIM_VERSION_100;

  dict_T *const msgpack_types_dict = tv_dict_alloc();
  for (size_t i = 0; i < ARRAY_SIZE(msgpack_type_names); i++) {
    list_T *const type_list = tv_list_alloc(0);
    tv_list_set_lock(type_list, VAR_FIXED);
    tv_list_ref(type_list);
    dictitem_T *const di = tv_dict_item_alloc(msgpack_type_names[i]);
    di->di_flags |= DI_FLAGS_RO|DI_FLAGS_FIX;
    di->di_tv = (typval_T) {
      .v_type = VAR_LIST,
      .vval = { .v_list = type_list, },
    };
    eval_msgpack_type_lists[i] = type_list;
    if (tv_dict_add(msgpack_types_dict, di) == FAIL) {
      // There must not be duplicate items in this dictionary by definition.
      abort();
    }
  }
  msgpack_types_dict->dv_lock = VAR_FIXED;

  set_vim_var_dict(VV_MSGPACK_TYPES, msgpack_types_dict);
  set_vim_var_dict(VV_COMPLETED_ITEM, tv_dict_alloc_lock(VAR_FIXED));

  set_vim_var_dict(VV_EVENT, tv_dict_alloc_lock(VAR_FIXED));
  set_vim_var_list(VV_ERRORS, tv_list_alloc(kListLenUnknown));
  set_vim_var_nr(VV_STDERR,   CHAN_STDERR);
  set_vim_var_nr(VV_SEARCHFORWARD, 1L);
  set_vim_var_nr(VV_HLSEARCH, 1L);
  set_vim_var_nr(VV_COUNT1, 1);
  set_vim_var_nr(VV_TYPE_NUMBER, VAR_TYPE_NUMBER);
  set_vim_var_nr(VV_TYPE_STRING, VAR_TYPE_STRING);
  set_vim_var_nr(VV_TYPE_FUNC,   VAR_TYPE_FUNC);
  set_vim_var_nr(VV_TYPE_LIST,   VAR_TYPE_LIST);
  set_vim_var_nr(VV_TYPE_DICT,   VAR_TYPE_DICT);
  set_vim_var_nr(VV_TYPE_FLOAT,  VAR_TYPE_FLOAT);
  set_vim_var_nr(VV_TYPE_BOOL,   VAR_TYPE_BOOL);
  set_vim_var_nr(VV_TYPE_BLOB,   VAR_TYPE_BLOB);

  set_vim_var_bool(VV_FALSE, kBoolVarFalse);
  set_vim_var_bool(VV_TRUE, kBoolVarTrue);
  set_vim_var_special(VV_NULL, kSpecialVarNull);
  set_vim_var_nr(VV_NUMBERMAX, VARNUMBER_MAX);
  set_vim_var_nr(VV_NUMBERMIN, VARNUMBER_MIN);
  set_vim_var_nr(VV_NUMBERSIZE, sizeof(varnumber_T) * 8);
  set_vim_var_special(VV_EXITING, kSpecialVarNull);

  set_vim_var_nr(VV_ECHOSPACE,    sc_col - 1);

  vimvars[VV_LUA].vv_type = VAR_PARTIAL;
  vvlua_partial = xcalloc(1, sizeof(partial_T));
  vimvars[VV_LUA].vv_partial = vvlua_partial;
  // this value shouldn't be printed, but if it is, do not crash
  vvlua_partial->pt_name = xmallocz(0);
  vvlua_partial->pt_refcount++;

  set_reg_var(0);  // default for v:register is not 0 but '"'
}
```

```c
assert(STRLEN(p->vv_name) <= 16);
```

16 was replaced with the macro based on this patch.

zeertzjq commented:
Yes, but `ChangedtickDictItem` is not the type of `vv_di`.

zeertzjq commented:
I decided to call it `VIMVAR_KEY_LEN`, as here in Nvim the equivalent of `dictitem16_T` is really only used in `vimvar`. `changedtick_di` doesn't use it. I didn't call it `VV_KEY_LEN` because Vim already has many macros starting with `VV_`.

zeertzjq commented:
`dictitem16_T` was used in two different places, and it was changed into two different things in commit <https://github.com/neovim/neovim/commit/fb146e80aa1ead96518f38b9684e39249bc83485>:

<https://github.com/neovim/neovim/commit/fb146e80aa1ead96518f38b9684e39249bc83485#diff-2d3b26fd2a7003950af527e242854587cfb54a0b4d8da113f26440cf4b6d279dR323>

<https://github.com/neovim/neovim/commit/fb146e80aa1ead96518f38b9684e39249bc83485#diff-93b1a06d1f521272b27866a6e720b14721d3290ec2d06eeae653c51da2419d21R434>

zeertzjq committed:
vim-patch:8.2.3549: mistakes in test comments

Problem:    Mistakes in test comments.
Solution:   Fix the comments. (closes vim/vim#9029)
https://github.com/vim/vim/commit/b811de5d49a23b20df82a80533d0602cda49029e

zeertzjq committed:
vim-patch:8.1.1542: an OptionSet autocommand does not get enough info

Problem:    An OptionSet autocommand does not get enough info.
Solution:   Add v:option_command, v:option_oldlocal and v:option_oldglobal.
            (Latrice Wilgus, closes vim/vim#4118)
https://github.com/vim/vim/commit/d7c968794710f338d491072171df48f96612cf72

janlazo commented on src/nvim/option.c:
```diff
@@ -1596,6 +1617,8 @@ int do_set(char_u *arg, int opt_flags)
             // did_set_string_option(), make a copy.
             saved_origval = (origval != NULL) ? xstrdup((char *)origval) : 0;
+            saved_origval_l = (origval_l != NULL) ? xstrdup((char *)origval_l) : 0;
+            saved_origval_g = (origval_g != NULL) ? xstrdup((char *)origval_g) : 0;
```
Why 0 instead of NULL? Compiler can replace `0`, `NULL` with the correct value but this variable is a pointer.

janlazo replied:
Nvrmnd. didn't notice the lines above. Still, that's confusing to me.
